extension/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod constants;
+#[cfg(test)]
+mod tests;
+mod v0;
+
+use codec::Encode;
+use constants::*;
+use frame_support::{
+	dispatch::{GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::{Contains, OriginTrait},
+};
+use frame_system::RawOrigin;
+use pallet_contracts::chain_extension::{
+	BufInBufOutState, ChainExtension, Environment, Ext, InitState, RetVal,
+};
+use sp_core::crypto::UncheckedFrom;
+use sp_runtime::{traits::Dispatchable, DispatchError};
+use sp_std::vec::Vec;
+
+type ContractSchedule<T> = <T as pallet_contracts::Config>::Schedule;
+
+pub trait Config:
+	frame_system::Config<RuntimeCall: GetDispatchInfo + Dispatchable<PostInfo = PostDispatchInfo>>
+{
+	/// A query of runtime state.
+	type RuntimeRead: Decode + ReadStateHandler<Self>;
+	/// Allowlisted runtime calls and read state calls.
+	type AllowedApiCalls: Contains<Self::RuntimeCall> + Contains<Self::RuntimeRead>;
+}
+
+/// Trait for handling parameters from the chain extension environment during state read operations.
+pub trait ReadStateHandler<T: Config> {
+	/// Processes the parameters needed to execute a call within the runtime environment.
+	/// Layout of `params`: [version, pallet_index, call_index, ...Vec<u8>].
+	fn handle_read(params: T::RuntimeRead) -> Vec<u8>;
+}
+
+#[derive(Default)]
+pub struct ApiExtension;
```

Comment:
Rather than being a type, would using a trait with some default function implementations not provide the same result but in a more flexible way? I gave the example of the XcmExecutor elsewhere, but perhaps an ApiExtension trait could inherit from ChainExtension<T>, add the additional function in ReadStateHandler, or inherit from that as well, and provide a single trait that can then be implemented for a type in the runtime.